lib/Target/AMDGCN/AMDGCNTranslation.cpp

```diff
@@ -0,0 +1,100 @@
+#include "triton/Target/AMDGCN/AMDGCNTranslation.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/ExecutionEngine/ExecutionEngine.h"
+#include "mlir/ExecutionEngine/OptUtils.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.h"
+#include "mlir/Target/LLVMIR/Export.h"
+#include "mlir/Target/LLVMIR/LLVMTranslationInterface.h"
+#include "triton/Target/LLVMIR/LLVMIRTranslation.h"
+
+#include "llvm/ExecutionEngine/ExecutionEngine.h"
+#include "llvm/ExecutionEngine/SectionMemoryManager.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/IRPrintingPasses.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/MC/TargetRegistry.h"
+#include "llvm/Support/CodeGen.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/SourceMgr.h"
+#include "llvm/Support/TargetSelect.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Target/TargetMachine.h"
+#include "llvm/Target/TargetOptions.h"
+#include "llvm/Transforms/Scalar.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+
+namespace triton {
+
+static void init_llvm() {
+  LLVMInitializeAMDGPUTargetInfo();
+  LLVMInitializeAMDGPUTarget();
+  LLVMInitializeAMDGPUTargetMC();
+  LLVMInitializeAMDGPUAsmPrinter();
+}
+
+static std::string llir_to_amdgcn(llvm::Module *module,
+                                  const std::string &_proc) {
+  init_llvm();
+
+  llvm::SmallVector<char, 0> buffer;
+  std::string triple = "amdgcn-amd-amdhsa";
+  std::string layout = "";
+  std::string features = "+sramecc,-xnack";
+  // verify and store llvm
+  llvm::legacy::PassManager pm;
+  pm.add(llvm::createVerifierPass());
+  pm.run(*module);
+  // create machine
+  module->setTargetTriple(triple);
+  std::string error;
+  auto target =
+      llvm::TargetRegistry::lookupTarget(module->getTargetTriple(), error);
+  llvm::TargetOptions opt;
+
+  opt.AllowFPOpFusion = llvm::FPOpFusion::Fast;
+  opt.UnsafeFPMath = false;
+  opt.NoInfsFPMath = false;
+  opt.NoNaNsFPMath = true;
+
+  llvm::TargetMachine *machine = target->createTargetMachine(
+      module->getTargetTriple(), _proc, features, opt, llvm::Reloc::PIC_,
+      llvm::None, llvm::CodeGenOpt::None);
+
+  // set data layout
+  if (layout.empty())
+    module->setDataLayout(machine->createDataLayout());
+  else
+    module->setDataLayout(layout);
+  // emit machine code
+  for (llvm::Function &f : module->functions()) {
```

Comment:
@B1tway this line is too relax. need to be changed.

@rsanthanam-amd will share some more info.

Reply:
@B1tway 

The AMDGPU_KERNEL calling convention should only be added to kernel functions.

The problem with inserting the call in this spot is that it will add the calling convention to all functions regardless of whether or not they are kernel functions (some are device functions for example).

The other issue is adding the call here happens after the LLVM verifier has done it's job so this mistake will escape the verification step (see line 52-53 in this file).

A better place to set the calling convention would during the TritonGPU --> LLVM IR generation phase.

Reply:
@B1tway 

I noticed that the call to add the AMDGPU_KERNEL calling convention is already present in the right place:

https://github.com/B1tway/triton/blob/B1tway/amdgcn-dump/lib/Target/LLVMIR/LLVMIRTranslation.cpp#L55

But it is disabled because:

https://github.com/B1tway/triton/blob/B1tway/amdgcn-dump/lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp#L259-L260

The nvvm kernel annotation metadata is disabled for ROCm so I think if you enable that, it would solve this problem (and remove the call to adding the AMDGPU_KERNEL calling convention from lib/target/AMDGCN/AMDGCNTranslation.cpp).

Reply:
@rsanthanam-amd 
Thank you for the detailed explanation. I will try to fix it in the near future.